dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java

```diff
@@ -41,38 +57,94 @@ public void setToIgnoreMetadataFields(Map<Integer, List<String>> toIgnoreMetadat
         this.toIgnoreMetadataFields = toIgnoreMetadataFields;
     }
 
-    public DiscoveryConfiguration getDiscoveryConfiguration(IndexableObject dso) {
+    public DiscoveryConfiguration getDiscoveryConfiguration(final Context context,
+                                                            IndexableObject dso) {
         String name;
         if (dso == null) {
             name = "default";
         } else if (dso instanceof IndexableDSpaceObject) {
-            name = ((IndexableDSpaceObject) dso).getIndexedObject().getHandle();
+            return getDiscoveryDSOConfiguration(context, ((IndexableDSpaceObject) dso).getIndexedObject());
         } else {
             name = dso.getUniqueIndexID();
         }
-
         return getDiscoveryConfiguration(name);
     }
 
+    /**
+     * Retrieve the discovery configuration for the provided DSO. When no direct match is found, the parent object will
+     * be checked until there is no parent left, in which case the "default" configuration will be returned.
+     * @param context   - The database context
+     * @param dso       - The DSpace object to retrieve the configuration for
+     * @return the discovery configuration for the provided DSO.
+     */
+    public DiscoveryConfiguration getDiscoveryDSOConfiguration(final Context context,
+                                                               DSpaceObject dso) {
+        // Fall back to default configuration
+        if (dso == null) {
+            return getDiscoveryConfiguration("default", false);
+        }
+
+        // Attempt to retrieve cached configuration by UUID
+        if (uuidMap.containsKey(dso.getID())) {
+            return uuidMap.get(dso.getID());
+        }
+
+        DiscoveryConfiguration configuration;
+
+        // Attempt to retrieve configuration by DSO handle
+        configuration = getDiscoveryConfiguration(dso.getHandle(), false);
+
+        if (configuration == null) {
+            // Recurse up the Comm/Coll hierarchy until a configuration is found
+            DSpaceObjectService<DSpaceObject> dSpaceObjectService =
+                ContentServiceFactory.getInstance().getDSpaceObjectService(dso);
+            DSpaceObject parentObject = null;
+            try {
+                parentObject = dSpaceObjectService.getParentObject(context, dso);
+            } catch (SQLException e) {
+                log.error(e);
+            }
+            configuration = getDiscoveryDSOConfiguration(context, parentObject);
+        }
+
+        // Cache the resulting configuration
+        uuidMap.put(dso.getID(), configuration);
```

Comment:
only collection and community uuid should be cached and only for a thread duration

Reply:
@abollini limiting the cache to a single thread would undermine the performance gain for e.g. regular REST requests.

I don't think this is really necessary if we limit the cache to Comm/Coll only, since this makes the structure much simpler
* Comm/Coll structure can only change ~ create/delete
* When a Comm/Coll is created, it will be added to the cache the first time its discovery config is resolved
* When a Comm/Coll is deleted, it will remain in the cache until Tomcat is restarted. However, we only ever retrieve configurations for _existing_ DSOs, so this is not a problem.